### ISSUE TITLE
import-tar incorrectly modifies modes, owner and datetime of some directories

### ISSUE THREAD

**Issue by CarlosCastaneda** [borgbackup/borg#7669]
On my raspberry pi compressed tar backups of different and distinct directories (tar -C dir1 -C dir2 ...) are made on a regular basis and after successfully finishing I'm creating an archive of this backup in a borg repository. Verifying these tar imports (by means of tar --compare against the corresponding "borg mounted" borg archive) it turned out that for some directories invalid modes, ownership and datetime values are stored. 

Here are some data to reveil the issue:

this is an excerpt of the listing of the original tar backup

```
$ egrep 'data/user/files_versions/MyMusic/MP3/Charlie Haden & Pat Metheny/Beyond The Missouri Sky \(Short Stories\)|data/user/files_versions/MyMusic/MP3/Ferrer, Ibrahim/Buena Vista Social Club Presents Ibrahim Ferrer|data/user/files_versions/MyMusic/MP3/lebensart music.one/a smooth breeze of decadent songs for modern people' nextcloud-bkp_20230622_1687395606.tar.gz.txt
drwxr-x--- www-data/www-data        0 2022-03-19 12:45 data/user/files_versions/MyMusic/MP3/Charlie Haden & Pat Metheny/Beyond The Missouri Sky (Short Stories)/
drwxr-x--- www-data/www-data        0 2022-03-19 12:45 data/user/files_versions/MyMusic/MP3/Ferrer, Ibrahim/Buena Vista Social Club Presents Ibrahim Ferrer/
drwxr-x--- www-data/www-data        0 2022-03-19 12:45 data/user/files_versions/MyMusic/MP3/lebensart music.one/a smooth breeze of decadent songs for modern people/

```
This lines result from a borg (version 1.1.16) backup of the original directories
 
```
$ sudo bash -c "cd mnt; ls -ald 'data/user/files_versions/MyMusic/MP3/Charlie Haden & Pat Metheny/Beyond The Missouri Sky (Short Stories)' 'data/user/files_versions/MyMusic/MP3/Ferrer, Ibrahim/Buena Vista Social Club Presents Ibrahim Ferrer' 'data/user/files_versions/MyMusic/MP3/lebensart music.one/a smooth breeze of decadent songs for modern people'"
drwxr-x--- 1 www-data www-data 0 Mar 19  2022 'data/user/files_versions/MyMusic/MP3/Charlie Haden & Pat Metheny/Beyond The Missouri Sky (Short Stories)'
drwxr-x--- 1 www-data www-data 0 Mar 19  2022 'data/user/files_versions/MyMusic/MP3/Ferrer, Ibrahim/Buena Vista Social Club Presents Ibrahim Ferrer'
drwxr-x--- 1 www-data www-data 0 Mar 19  2022 'data/user/files_versions/MyMusic/MP3/lebensart music.one/a smooth breeze of decadent songs for modern people'

```
This lines result from the borg (version 1.2.4) import-tar (incorrect data: drwxr-xr-**x 1 root root 0 Jun 23 19:58**)
 
```
$ sudo bash -c "cd /tmp/mnt; ls -ald 'data/user/files_versions/MyMusic/MP3/Charlie Haden & Pat Metheny/Beyond The Missouri Sky (Short Stories)' 'data/user/files_versions/MyMusic/MP3/Ferrer, Ibrahim/Buena Vista Social Club Presents Ibrahim Ferrer' 'data/user/files_versions/MyMusic/MP3/lebensart music.one/a smooth breeze of decadent songs for modern people'"
drwxr-xr-x 1 root root 0 Jun 23 19:58 'data/user/files_versions/MyMusic/MP3/Charlie Haden & Pat Metheny/Beyond The Missouri Sky (Short Stories)'
drwxr-xr-x 1 root root 0 Jun 23 19:58 'data/user/files_versions/MyMusic/MP3/Ferrer, Ibrahim/Buena Vista Social Club Presents Ibrahim Ferrer'
drwxr-xr-x 1 root root 0 Jun 23 19:58 'data/user/files_versions/MyMusic/MP3/lebensart music.one/a smooth breeze of decadent songs for modern people'

```
Some system and version infos (borgneu is standalone and borg is the latest debian package version):

```
$ apt list --installed borg*
Listing... Done
borgbackup/oldstable,now 1.1.16-3 arm64 [installed]
$ borg --version
borg 1.1.16
$ borgneu --version
borgneu 1.2.4
$ uname -a
Linux nextcloudpi 6.1.21-v8+ #1642 SMP PREEMPT Mon Apr  3 17:24:16 BST 2023 aarch64 GNU/Linux
$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye
```

**Comment by ThomasWaldmann** [comment]
You didn't give the borg commands used.

Also, it would be good if you write a small and as-simple-as-it-gets reproducer script:
- use short, relative paths in your home (not interested in your music)
- use `tar`, `borg import-tar` and `borg list` to prove there is an issue with `borg import-tar` (we are not debugging your fuse configuration or usage)

**Comment by CarlosCastaneda** [comment]
> You didn't give the borg commands used.
> 
That's right, I should have done that right away:
`Command line: borgneu import-tar --timestamp nextcloud-bkp_20230622_1687395606.tar.gz --list --stats ::nextcloud-bkp_20230622_1687395606.tar.gz nextcloud-bkp_20230622_1687395606.tar.gz`
> Also, it would be good if you write a small and as-simple-as-it-gets reproducer script:
> 
>     * use borg 1.2.4
Hence, I will uninstall the package version and use the latest standalone version from now on.
> 
>     * use short, relative paths in your home (not interested in your music)
These are my data and the mentioned directories (music) were already the shortest of them. 
> 
>     * use `tar`, `borg import-tar` and `borg list` to prove there is an issue with `borg import-tar` (we are not debugging your fuse configuration or usage)

**borg list** was the ultimate hint here. Thanks. Using the borg list command everything seems to be correct, at least I don't see any differences in the listing.

Actually I just wanted to implement a simple method to check my import with borg import-tar and thought that a borg mounted file system would fit well along with tar --compare. I didn't want to rely on a listing only but also check the data.

However, the fact is then that with borg mount some directories show a different owner, mode and datetime value. This is not consistently the case everywhere, but apparently only for (some) directories (at 7762 among 53823) and not for a single file. That's what I've been able to determine so far (with 82 import-tars), but I have no idea or any clue as to what that might be reason.

Then it is more likely a problem in the interaction of borg mount and filesystem in userspace. Since I have not configured anything in either, I would expect consistent behavior. E.g. either all files and directories belong to root or to the real owner (at the time of backup). By the way, this is all running on the same machine (in terms of ownership and user IDs).

**Comment by ThomasWaldmann** [comment]
BTW, how `borg mount` builds the fs is influenced by `--numeric-ids` (and the default for that changed at some borg version to behave more similar to `borg extract`. And, depending on the system you used to create the archive and to look at the mounted fs, the uid/gid <-> user/group-name mapping might be different.

mode: no idea right now why it could be different.

timestamps: note that borg archives 2 timestamps: ctime and mtime (optionally it can also archive atime, but that is not too useful usually).

I guess I could do more debugging if you could write a script that reproduces an issue (so I can run it here).

**Comment by CarlosCastaneda** [comment]
> BTW, how `borg mount` builds the fs is influenced by `--numeric-ids` (and the default for that changed at some borg version to behave more similar to `borg extract`. And, depending on the system you used to create the archive and to look at the mounted fs, the uid/gid <-> user/group-name mapping might be different.
> 
In my case there should be no difference as all is run on very same box
> mode: no idea right now why it could be different.
> 
> timestamps: note that borg archives 2 timestamps: ctime and mtime (optionally it can also archive atime, but that is not too useful usually).
With regard to timestamps, I'd suggest to be able to choose the creation time of the tar file using --timstamp parameter. Today the timestamp is ctime or mtime.
> 
> I guess I could do more debugging if you could write a script that reproduces an issue (so I can run it here).
I've setup a repo and stripped it down to a size of 33MB containing tar file to import as well as correspondig borg repository and some listing to demonstrate the difference. Is there a possibility to upload this file?

**Comment by CarlosCastaneda** [comment]
Repro:

```
# mytar is location of tar appended

mytar=./borg_test2.tar.gz
borg_repo_dir=$(mktemp -d)
borg_mount_dir=$(mktemp -d)

```
```
# set up repo and import tar file

sudo borg init --encryption repokey $borg_repo_dir
sudo borg import-tar --timestamp $mytar --list --stats $borg_repo_dir::borg_test2.tar.gz $mytar
```

```
# mount tar archive
sudo -E borg mount $borg_repo_dir::borg_test2.tar.gz $borg_mount_dir
```

```
# check for files with differnt owner to www-data
sudo find $borg_mount_dir ! -user www-data
# sudo find $borg_mount_dir  -user www-data

```
```
# create user listing

sudo bash -c "tar tvzf borg_test2.tar.gz > borg_test2.tar.gz.lis"
sudo bash -c "borg list $borg_repo_dir::borg_test2.tar.gz > borg_repo_dir__borg_test2.tar.gz.lis"
sudo bash -c "find $borg_mount_dir -print0 | xargs -0 sudo ls -ald > borg_mount_dir__borg_test2.tar.gz.lis"
```

```
# check counts = www-data as well as <> www-data.
# Should be the same in all files
grep -c 'www-data[/ ]www-data' *t2*.lis
borg_mount_dir__borg_test2.tar.gz.lis:112
borg_repo_dir__borg_test2.tar.gz.lis:117
borg_test2.tar.gz.lis:117

grep -vc 'www-data[/ ]www-data' *t2*.lis
borg_mount_dir__borg_test2.tar.gz.lis:11
borg_repo_dir__borg_test2.tar.gz.lis:0
borg_test2.tar.gz.lis:0
```
[borg_test2.tar.gz](https://github.com/borgbackup/borg/files/11860823/borg_test2.tar.gz)

**Comment by ThomasWaldmann** [comment]
does anything of what you do need `sudo`?

**Comment by CarlosCastaneda** [comment]
> does anything of what you do need `sudo`?

I'm not entirely sure, but I don't think so. It's due to different ownerships of files and file systems on my box. And with regard to the repro I tried to be as close as possible to the production environment.

**Comment by ThomasWaldmann** [comment]
Well, if you start from a tar file, you could do all that without sudo (including the `borg mount`), always using the same user.

So, try to reproduce your issue without sudo, please.

**Comment by CarlosCastaneda** [comment]
Yes, it is reproducible. But, because of missing permissions I had to change BORG_BASE_DIR to be able set up a new and independant repository in /tmp. Importing tar and mount works, but some files are now (incorrectly owned) by the current user (previously root), modes modified as well as datetime, which seem to be  the current time.

The current user is not a privileged user for security reason. So for mapping the uid/guid sudo seems to be the correct method after all. And sudo seems to take into account this description of mount in the documentation. Using sudo I'd expect to have the very same layout as has been backed up according to:

> Borg’s default behavior is to use the archived user and group names of each
> file and map them to the system’s respective user and group ids.
which are the same on the same box.

Here are the results of some commands:

```
# First run, permission issue
borg init --encryption repokey $borg_repo_dir
Error: [Errno 13] Permission denied: '/home/piuser/.config/borg/security'
```
After successfull import-tar and mount...
```
find $borg_mount_dir ! -user www-data
/tmp/borg_mount
/tmp/borg_mount/data
/tmp/borg_mount/data/updater-oclaqn5ofjmm
/tmp/borg_mount/data/updater-oclaqn5ofjmm/backups
/tmp/borg_mount/data/updater-oclaqn5ofjmm/backups/nextcloud-25.0.6.1-1682067142
/tmp/borg_mount/data/updater-oclaqn5ofjmm/backups/nextcloud-25.0.6.1-1682067142/3rdparty
find: ‘/tmp/borg_mount/data/updater-oclaqn5ofjmm/backups/nextcloud-25.0.6.1-1682067142/3rdparty/guzzlehttp’: Permission denied
```

```
find $borg_mount_dir -user www-data
/tmp/borg_mount/data/updater-oclaqn5ofjmm/backups/nextcloud-25.0.6.1-1682067142/3rdparty/guzzlehttp
find: ‘/tmp/borg_mount/data/updater-oclaqn5ofjmm/backups/nextcloud-25.0.6.1-1682067142/3rdparty/guzzlehttp’: Permission denied
```

```
sudo find $borg_mount_dir -user www-data
find: ‘/tmp/borg_mount’: Permission denied
```

```
find $borg_mount_dir ! -user www-data | xargs ls -al
find: ‘/tmp/borg_mount/data/updater-oclaqn5ofjmm/backups/nextcloud-25.0.6.1-1682067142/3rdparty/guzzlehttp’: Permission denied
/tmp/borg_mount:
total 4
drwxr-xr-x  1 piuser pi      0 Jun 26 16:46 .
drwxrwxrwt 16 root    root 4096 Jun 26 16:49 ..
drwxr-xr-x  1 piuser pi      0 Jun 26 16:46 data

/tmp/borg_mount/data:
total 0
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 .
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 ..
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 updater-oclaqn5ofjmm

/tmp/borg_mount/data/updater-oclaqn5ofjmm:
total 0
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 .
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 ..
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 backups

/tmp/borg_mount/data/updater-oclaqn5ofjmm/backups:
total 0
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 .
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 ..
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 nextcloud-25.0.6.1-1682067142

/tmp/borg_mount/data/updater-oclaqn5ofjmm/backups/nextcloud-25.0.6.1-1682067142:
total 0
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 .
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 ..
drwxr-xr-x 1 piuser pi 0 Jun 26 16:46 3rdparty

/tmp/borg_mount/data/updater-oclaqn5ofjmm/backups/nextcloud-25.0.6.1-1682067142/3rdparty:
total 0
drwxr-xr-x 1 piuser  pi       0 Jun 26 16:46 .
drwxr-xr-x 1 piuser  pi       0 Jun 26 16:46 ..
drwxr-x--- 1 www-data www-data 0 Apr 21 10:53 guzzlehttp

```
`borg list` itself returns correct data, modes user ownership and datetime.

**Comment by CarlosCastaneda** [comment]
In my opinion I have provided a perfect repro, little data (136KB of 36GB data, 117 entries out of 191737 entries) along with all statements necesseary to setup an reprduce the issue. Everything is included to analyze the problem. You have the software, a debugger and the perfect know how of your software. It's easy to set a watchpoint or breakpoint in debugger at the appropriate software code and see what's going (wrong). For me it's all a black box and guessing what the problem could be. I'd say it's your turn now.

**Comment by CarlosCastaneda** [comment]
I've succeded to reduce the repro even further with a single file in a directory. The borg commands mentioned above remain the same. New data file is appended.
[borg_test.tar.gz](https://github.com/borgbackup/borg/files/11904480/borg_test.tar.gz)

**Comment by ThomasWaldmann** [comment]
Created a borg `repo`, imported your tar - everything done as normal user "tw" except the `ls`.

tar listing:
```
drwxr-x--- www-data/www-data 0 2023-04-21 08:52 private/Collaboration/AutoComplete/
-rw-r--r-- www-data/www-data 2608 2023-04-21 08:52 private/Collaboration/AutoComplete/Manager.php
```

```
# note:
# to be able to access the mount as another user (here: root),
# we need to give mount option "allow_other".
# to be allowed to do this, we need "user_allow_other" in /etc/fuse.conf .
$ borg mount -o allow_other repo mnt

# we need root permissions here, otherwise it does not recurse into "AutoComplete" dir
$ sudo ls -lR mnt
mnt:
drwxr-xr-x 1 tw tw 0 Jun 29 13:46 arch

mnt/arch:
drwxr-xr-x 1 tw tw 0 Jun 29 13:50 private

mnt/arch/private:
drwxr-xr-x 1 tw tw 0 Jun 29 13:50 Collaboration

mnt/arch/private/Collaboration:
drwxr-x--- 1 www-data www-data 0 Apr 21 08:52 AutoComplete

mnt/arch/private/Collaboration/AutoComplete:
-rw-r--r-- 1 www-data www-data 2608 Apr 21 08:52 Manager.php
```

User, group and mode of all archived files (`Manager.php`) and dirs (`AutoComplete`) are correct in the FUSE fs.

Note that you did **not** archive all parent dir fs items and thus user/group/mode of these are just according to the current user/group and umask. We already have at least one issue about this on the issue tracker. Is it that what you meant?

**Comment by ThomasWaldmann** [comment]
Closing this:
- wrong usage of FUSE maybe?
- parent dirs not archived, we already have that, duplicate.

**Comment by CarlosCastaneda** [comment]
> Closing this:
> 
>     * wrong usage of FUSE maybe?
Thanks for your hints, I was not aware of this.
> 
>     * parent dirs not archived / not extracted, we already have an issue about that, duplicate.
Missing parents are due to the reqirements of the repro: reproduce in a simple way the behaviour with few files/directories and little data.

I have taken a closer look at my backups.

At the moment I run a weekly backup which saves a database backup (db.tar.gz) and 2 directories (data, nextcloud) in another compressed tar file. Since borg doesn't support the tar parameter "-C" I'm currently backing up the two directories into 2 distinct archives in the same repository on a daily basis. Additionally I imported the old tar backups and the current weekly ones into another borg repository. These tar imports I check for completeness (borg mount along with tar --compare).

Now I have noticed the following:
a borg list Repo1::data and Repo1::nextcloud is exactly the same with borg list Repo2::TarArchive, which was to be expected (with respect to user, modes, and datetime).
borg mount with Repo1::data and Repo::nextcloud shows that everything is correct (user, modes, datetime).
In contrast, borg mount Repo2::TarArchive shows the behavior described here. Since the two directory trees data and nextcloud are (tar) backed up directly (-C to parent directory) there are no parent directories to be backuped. From the logic Repo1 archives and the Repo2 tar archive are very the same. The only difference is borg create versus borg import-tar.
